Bugzilla/Config.pm

```diff
@@ -11,375 +11,366 @@ use 5.10.1;
   }
 
-  Bugzilla->params->{$name} = $value;
+  $self->{params}->{$name} = $value;
 }
 
-sub update_params {
-  my ($params) = @_;
-  my $answer = Bugzilla->installation_answers;
+sub update {
+  my ($self, $params) = @_;
+  $params ||= $self->{params};
 
-  my $param = read_param_file();
-  my %new_params;
+  try {
+    my $dbh = Bugzilla->dbh;
+    foreach my $key (keys %{$params}) {
+      my $new_value = $params->{$key} || '';
+      if ($dbh->selectrow_array('SELECT 1 FROM params WHERE name = ?', undef, $key)) {
+        $dbh->do('UPDATE params SET value = ? WHERE name = ?', undef, $new_value, $key);
+      }
+      else {
+        $dbh->do('INSERT INTO params (name, value) VALUES (?, ?)',
+          undef, $key, $new_value);
+      }
+    }
+
+    # And now we have to reset the params cache
+    Bugzilla->memcached->set_params($params);
+    delete Bugzilla->request_cache->{params};
+    delete Bugzilla->request_cache->{params_obj};
+    $self->{params} = $params;
+  }
+  catch {
+    WARN("Database not yet available: $_");
+  };
+}
+
+sub migrate_params {
+  my ($self) = @_;
+  my $answers = Bugzilla->installation_answers;
+  my $params  = $self->{params};
+
+  # Migrate old file based parameters to the database
+
+  my $datadir = bz_locations()->{'datadir'};
+  if (-e "$datadir/params") {
+
+    # Read in the old data/params values
+    my $s = Safe->new;
+    $s->rdo("$datadir/params");
+    die "Error reading $datadir/params: $!"    if $!;
+    die "Error evaluating $datadir/params: $@" if $@;
+    my %file_params = %{$s->varglob('param')};
+
+    WARN('Migrating old parameters from data/params to database');
+    foreach my $key (keys %file_params) {
+      $params->{$key} = $file_params{$key};
+    }
+
+    WARN('Backing up old params file');
+    rename("$datadir/params", "$datadir/params.old")
+      or die "Rename params file failed: $!";
+  }
 
-  # If we didn't return any param values, then this is a new installation.
-  my $new_install = !(keys %$param);
+  # Update old params
 
-  # --- UPDATE OLD PARAMS ---
+  my %new_params;
 
   # Change from usebrowserinfo to defaultplatform/defaultopsys combo
-  if (exists $param->{'usebrowserinfo'}) {
-    if (!$param->{'usebrowserinfo'}) {
-      if (!exists $param->{'defaultplatform'}) {
+  if (exists $params->{'usebrowserinfo'}) {
+    if (!$params->{'usebrowserinfo'}) {
+      if (!exists $params->{'defaultplatform'}) {
         $new_params{'defaultplatform'} = 'Other';
       }
-      if (!exists $param->{'defaultopsys'}) {
+      if (!exists $params->{'defaultopsys'}) {
         $new_params{'defaultopsys'} = 'Other';
       }
     }
   }
 
   # Change from a boolean for quips to multi-state
-  if (exists $param->{'usequip'} && !exists $param->{'enablequips'}) {
-    $new_params{'enablequips'} = $param->{'usequip'} ? 'on' : 'off';
+  if (exists $params->{'usequip'}
+    && !exists $params->{'enablequips'})
+  {
+    $new_params{'enablequips'} = $params->{'usequip'} ? 'on' : 'off';
```

Comment:
Oof, that `$param` -> `$params` rename hurts a little in making this diff a little more aggressive.
If you don't mind splitting such impactful changes into separate commits, that could make reviewing a little bit easier.